### DISCUSSION TITLE
Rust re-implementation has 2x Performance Boost!

### DISCUSSION THREAD

**Discussion by HoKim98** [PRBonn/kiss-icp#273] [Ideas]
Hello, thank you to all the community contributors who are leading this great project.

After porting the existing core package written in `C++` to `Rust`, I successfully achieved performance that was 20%~100% better than the existing binary by applying various modern multi-threading optimization techniques. In addition, I have developed a fully compatible `pyo3` library (Rust version of `pybind`) so that it can seamlessly replace the existing pybind library.

However, I cannot publish the source code and binaries, because of `Licensing`. I translated and modified the C++ source code logic under the MIT license, but I am actively introducing various functions and abstractions for compatibility with my personal projects, so I feel the burden of merging it directly into the original repository. I would like to open the source codes to the community and also I want to freely edit the Rust source code. Currently, I am writing it in a new repository and plan to use the original license file as is.

If you can afford it, please reply below this discussion, or contact me via e-mail!

**Comment by nachovizzo** [comment]
@kerryeon thanks for reaching out. It's truly gratifying to see your interest in our library and all the effort you are making to porting to `rust`

Honestly I wanted to do this myself as an exercise of learning `rust` but never found time.

In terms of the LICENSE, I don't fully understand your concern. Should we try to loosen our license? Or would you like to contribute the rust implementation to the mainstream repo? Honestly, I'd be interested in the 2nd option, I believe `kiss-icp` already serves (or that was always my intention) as an example package for a robotics library with ROS applications (but not the other way around). I would love to see the `rust` folder right next to the `cpp` and the `python` one. Similar to the mechanism we have to build the same source code for ROS 1 / ROS 2 we could have some switches to select between the C++/Rust backend (I believe).

Happy to open a discussion here with the other mantainers of the project as well. @tizianoGuadagnino , @benemer what do you guys think?

**Reply by benemer** [reply]
I agree that it would be super cool to also have a `rust` backend! Thanks for all this effort :)

I am just a bit concerned that this complicates the build system and code readability because for now, we have a `cpp` core which is used with `cpp`, `python`, ROS 1, and ROS 2 "wrappers". This change would mean having the same functionalities duplicated in two cores which both have to be maintained. I am not sure if we can maintain the `rust` core appropriately, at least I have zero experience with `rust` (yet).
